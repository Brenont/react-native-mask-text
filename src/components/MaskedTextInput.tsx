import React, { useEffect, useState } from "react";
import { TextInput, TextInputProps } from "react-native";
import { mask, unMask } from "../utils/mask";

interface OnChangeTextProps {
  text: String;
  rawText: String;
}

interface MaskedTextInputProps {
  mask: string;
  onChangeText: (
    maskedValue: string,
    unMaskedValue: string
  ) => OnChangeTextProps;
}

export default function MaskedTextInput({
  mask: pattern,
  onChangeText,
  ...rest
}: MaskedTextInputProps & TextInputProps) {
  const [maskedValue, setMaskedValue] = useState("");
  const [unMaskedValue, setUnmaskedValue] = useState("");

  function onChange(value: string) {
    const newMaskedValue = mask(value, pattern);

    setMaskedValue(newMaskedValue);
    setUnmaskedValue(unMask(newMaskedValue));
  }

  useEffect(() => {
    onChangeText(maskedValue, unMaskedValue);
  }, [maskedValue, unMaskedValue]);

  return (
    <TextInput
      onChangeText={(value) => onChange(value)}
      value={maskedValue}
      {...rest}
    />
  );
}
