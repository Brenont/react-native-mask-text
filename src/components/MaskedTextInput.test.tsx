import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { MaskedTextInput } from './MaskedTextInput';

describe('<MaskedTextInput />', () => {
  const mockedOnChangeText = jest.fn();

  test('should renders correctly', () => {
    const container = render(<MaskedTextInput mask="AAA-999" onChangeText={mockedOnChangeText} />);
    expect(container).toMatchSnapshot();
  });

  test('should mask input text', async () => {
    const container = render(
      <MaskedTextInput mask="AAA-999" onChangeText={mockedOnChangeText} testID="masked-text-input"/>
    );

    fireEvent.changeText(container.getByTestId("masked-text-input"), 'RCT777')

    await waitFor(() => {
      expect(mockedOnChangeText).toHaveBeenCalledTimes(3);
      expect(container.getByDisplayValue("RCT-777")).toBeTruthy();
    })
  });
})