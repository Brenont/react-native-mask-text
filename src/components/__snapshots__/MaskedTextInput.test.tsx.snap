// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MaskedTextInput /> should renders correctly 1`] = `
<TextInput
  allowFontScaling={true}
  onChangeText={[Function]}
  rejectResponderTermination={true}
  underlineColorAndroid="transparent"
  value=""
/>
`;
